File: android/app/src/main/kotlin/com/niu/flyguy/camera/AspenCameraHandler.kt
package com.niu.flyguy.camera

import android.app.Activity
import android.util.Log
import com.jovision.jvplayer.callback.JVSRequestCallback
import com.jovision.jvplayer.data.response.EmptyBean
import com.jovision.jvplayer.player.JVPlayerUtil
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Aspen 摄像头通信处理器
 *
 * 职责：
 * 1. MethodChannel: 处理 Flutter 的控制调用（play, pause, dispose 等）
 * 2. EventChannel: 推送播放器状态事件到 Flutter
 */
class AspenCameraHandler(
    private val flutterEngine: FlutterEngine,
    private val activity: Activity
) {
    companion object {
        private const val TAG = "AspenCameraHandler"

        // MethodChannel: Flutter 调用 Android
        private const val METHOD_CHANNEL_NAME = "camera/aspen"

        // EventChannel: Android 推送事件到 Flutter
        private const val EVENT_CHANNEL_NAME = "camera/aspen/events"
    }

    // ═══════════════════════════════════════════════════════════════
    // MethodChannel - Flutter 调用 Android 的控制方法
    // ═══════════════════════════════════════════════════════════════
    private val methodChannel: MethodChannel = MethodChannel(
        flutterEngine.dartExecutor.binaryMessenger,
        METHOD_CHANNEL_NAME
    )

    // ═══════════════════════════════════════════════════════════════
    // EventChannel - Android 主动推送事件到 Flutter
    // ═══════════════════════════════════════════════════════════════
    private val eventChannel: EventChannel = EventChannel(
        flutterEngine.dartExecutor.binaryMessenger,
        EVENT_CHANNEL_NAME
    )

    // EventChannel 的事件发送器（Flutter 监听时会设置）
    private var eventSink: EventChannel.EventSink? = null

    private var isPlayerInitialized = false

    fun register() {
        Log.d(TAG, "注册通信通道")

        // ─────────────────────────────────────────────────────────
        // 注册 MethodChannel（处理 Flutter 的方法调用）
        // ─────────────────────────────────────────────────────────
        methodChannel.setMethodCallHandler { call, result ->
            try {
                when (call.method) {
                    "initialize" -> handleInitialize(call, result)
                    "play" -> handlePlay(result)
                    "pause" -> handlePause(result)
                    "dispose" -> handleDispose(result)
                    "takeSnapshot" -> handleSnapshot(result)
                    "switchQuality" -> handleSwitchQuality(call, result)
                    else -> result.notImplemented()
                }
            } catch (e: Exception) {
                Log.e(TAG, "MethodChannel 方法调用异常: ${call.method}", e)
                result.error("EXCEPTION", e.message, e.stackTraceToString())
            }
        }

        // ─────────────────────────────────────────────────────────
        // 注册 EventChannel（监听 Flutter 的订阅）
        // ─────────────────────────────────────────────────────────
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            // Flutter 开始监听时调用
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                Log.d(TAG, "EventChannel: Flutter 开始监听播放器事件")
                eventSink = events
            }

            // Flutter 停止监听时调用
            override fun onCancel(arguments: Any?) {
                Log.d(TAG, "EventChannel: Flutter 停止监听播放器事件")
                eventSink = null
            }
        })

        Log.d(TAG, "通信通道注册完成")
    }

    // ═══════════════════════════════════════════════════════════════
    // EventChannel - 发送事件到 Flutter 的公共方法
    // ═══════════════════════════════════════════════════════════════

    /**
     * 发送播放器事件到 Flutter
     *
     * @param event 事件数据（Map 格式）
     *
     * 示例：
     * sendPlayerEvent(mapOf(
     *     "type" to "player",
     *     "event" to "video_ready"
     * ))
     */
    fun sendPlayerEvent(event: Map<String, Any>) {
        activity.runOnUiThread {
            eventSink?.success(event)
            Log.d(TAG, "EventChannel: 发送事件到 Flutter: $event")
        }
    }

    /**
     * 发送错误到 Flutter
     *
     * @param code 错误代码
     * @param message 错误消息
     * @param details 详细信息（可选）
     */
    fun sendPlayerError(code: String, message: String?, details: Any? = null) {
        activity.runOnUiThread {
            eventSink?.error(code, message, details)
            Log.e(TAG, "EventChannel: 发送错误到 Flutter: $code - $message")
        }
    }

    // ═══════════════════════════════════════════════════════════════
    // MethodChannel - 处理 Flutter 的方法调用
    // ═══════════════════════════════════════════════════════════════

    private fun handleInitialize(call: MethodCall, result: MethodChannel.Result) {
        val deviceId = call.argument<String>("deviceId")
        Log.d(TAG, "MethodChannel: 初始化播放器: deviceId=$deviceId")

        if (deviceId.isNullOrEmpty()) {
            result.error("INVALID_PARAMETER", "deviceId 不能为空", null)
            return
        }

        try {
            isPlayerInitialized = true
            result.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "MethodChannel: 初始化失败", e)
            result.error("INIT_ERROR", e.message, e.stackTraceToString())
        }
    }

    private fun handlePlay(result: MethodChannel.Result) {
        Log.d(TAG, "MethodChannel: 开始播放")
        try {
            JVPlayerUtil.instance.initP2P()
            JVPlayerUtil.instance.onResume()
            JVPlayerUtil.instance.startPlayer()
            result.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "MethodChannel: 播放失败", e)
            result.error("PLAY_ERROR", e.message, e.stackTraceToString())
        }
    }

    private fun handlePause(result: MethodChannel.Result) {
        Log.d(TAG, "MethodChannel: 暂停播放")
        try {
            JVPlayerUtil.instance.stopPlayer()
            JVPlayerUtil.instance.onDestroy()
            result.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "MethodChannel: 暂停失败", e)
            result.error("PAUSE_ERROR", e.message, e.stackTraceToString())
        }
    }

    private fun handleDispose(result: MethodChannel.Result) {
        Log.d(TAG, "MethodChannel: 释放播放器资源")
        try {
            JVPlayerUtil.instance.stopPlayer()
            JVPlayerUtil.instance.onDestroy()
            isPlayerInitialized = false
            result.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "MethodChannel: 资源释放失败", e)
            result.error("DISPOSE_ERROR", e.message, e.stackTraceToString())
        }
    }

    private fun handleSnapshot(result: MethodChannel.Result) {
        Log.d(TAG, "MethodChannel: 截图")
        try {
            val fileName = "snapshot_${System.currentTimeMillis()}.jpg"
            val filePath = activity.getExternalFilesDir("Pictures")?.absolutePath + "/$fileName"
            JVPlayerUtil.instance.startSnapshot(1, filePath)
            result.success(filePath)
        } catch (e: Exception) {
            Log.e(TAG, "MethodChannel: 截图失败", e)
            result.error("SNAPSHOT_ERROR", e.message, e.stackTraceToString())
        }
    }

    private fun handleSwitchQuality(call: MethodCall, result: MethodChannel.Result) {
        val quality = call.argument<String>("quality")
        Log.d(TAG, "MethodChannel: 切换清晰度: $quality")

        try {
            val streamType = if (quality == "low") 1 else 0
            JVPlayerUtil.instance.switchStreamType(streamType, object : JVSRequestCallback<EmptyBean> {
                override fun onSuccess(emptyBean: EmptyBean) {
                    result.success(null)
                }

                override fun onFailure(code: Int) {
                    result.error("QUALITY_ERROR", "切换失败: $code", null)
                }
            })
        } catch (e: Exception) {
            result.error("QUALITY_ERROR", e.message, e.stackTraceToString())
        }
    }

    fun unregister() {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }
}

File: ios/Runner/AspenCameraHandler.swift
import Flutter
import UIKit

/// Aspen 摄像头通信处理器
///
/// 职责：
/// 1. MethodChannel: 处理 Flutter 的控制调用（play, pause, dispose 等）
/// 2. EventChannel: 推送播放器状态事件到 Flutter
class AspenCameraHandler: NSObject {

    // ═══════════════════════════════════════════════════════════════
    // MethodChannel - Flutter 调用 iOS
    // ═══════════════════════════════════════════════════════════════
    private let methodChannel: FlutterMethodChannel

    // ═══════════════════════════════════════════════════════════════
    // EventChannel - iOS 推送事件到 Flutter
    // ═══════════════════════════════════════════════════════════════
    private let eventChannel: FlutterEventChannel
    private var eventSink: FlutterEventSink?

    // SDK 相关
    private var isPlayerInitialized = false

    // 通道名称常量
    private static let METHOD_CHANNEL_NAME = "camera/aspen"
    private static let EVENT_CHANNEL_NAME = "camera/aspen/events"

    init(messenger: FlutterBinaryMessenger) {
        self.methodChannel = FlutterMethodChannel(
            name: AspenCameraHandler.METHOD_CHANNEL_NAME,
            binaryMessenger: messenger
        )

        self.eventChannel = FlutterEventChannel(
            name: AspenCameraHandler.EVENT_CHANNEL_NAME,
            binaryMessenger: messenger
        )

        super.init()

        print("AspenCameraHandler: 初始化完成")
    }

    // ─────────────────────────────────────────────────────────
    // 注册通道
    // ─────────────────────────────────────────────────────────
    func register() {
        print("AspenCameraHandler: 注册通信通道")

        // 注册 MethodChannel
        methodChannel.setMethodCallHandler { [weak self] (call, result) in
            self?.handleMethodCall(call: call, result: result)
        }

        // 注册 EventChannel
        eventChannel.setStreamHandler(self)
    }

    // ─────────────────────────────────────────────────────────
    // 处理 Flutter 的方法调用
    // ─────────────────────────────────────────────────────────
    private func handleMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        do {
            switch call.method {
            case "initialize":
                handleInitialize(call: call, result: result)
            case "play":
                handlePlay(result: result)
            case "pause":
                handlePause(result: result)
            case "dispose":
                handleDispose(result: result)
            case "takeSnapshot":
                handleSnapshot(result: result)
            case "switchQuality":
                handleSwitchQuality(call: call, result: result)
            default:
                result(FlutterMethodNotImplemented)
            }
        } catch {
            print("AspenCameraHandler: MethodChannel 方法调用异常: \(call.method), \(error)")
            result(FlutterError(
                code: "EXCEPTION",
                message: error.localizedDescription,
                details: nil
            ))
        }
    }

    // ═══════════════════════════════════════════════════════════════
    // MethodChannel 方法实现
    // ═══════════════════════════════════════════════════════════════

    private func handleInitialize(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let deviceId = args["deviceId"] as? String,
              let p2pId = args["p2pId"] as? String,
              let p2pInitString = args["p2pInitString"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGS",
                message: "初始化参数缺失",
                details: nil
            ))
            return
        }

        let channelNo = args["channelNo"] as? Int ?? 0

        print("AspenCameraHandler: MethodChannel 初始化")
        print("  deviceId=\(deviceId)")
        print("  p2pId=\(p2pId)")
        print("  channelNo=\(channelNo)")

        // iOS SDK 不需要在这里初始化，会在 PlatformView 中初始化
        // 这里只是记录状态
        isPlayerInitialized = true
        result(nil)
    }

    private func handlePlay(result: @escaping FlutterResult) {
        print("AspenCameraHandler: MethodChannel 开始播放")

        // iOS 端播放逻辑由 PlatformView 控制
        // 这里预留接口，可以通过通知或代理模式通知 PlatformView

        result(nil)
    }

    private func handlePause(result: @escaping FlutterResult) {
        print("AspenCameraHandler: MethodChannel 暂停播放")

        // iOS 端暂停逻辑由 PlatformView 控制

        result(nil)
    }

    private func handleDispose(result: @escaping FlutterResult) {
        print("AspenCameraHandler: MethodChannel 收到 dispose 请求")

        // ✅ 停止并清理当前激活的播放器（等同于 Android 的 stopPlayer + onDestroy）
        AspenCameraViewFactory.viewLock.lock()

        print("AspenCameraHandler: 检查 activeView 状态...")
        if let activeView = AspenCameraViewFactory.activeView {
            print("AspenCameraHandler: ✅ 找到激活的播放器 viewId=\(activeView.viewId)")
            activeView.forceCleanup()
            print("AspenCameraHandler: ✅ 播放器已停止并清理")
        } else {
            print("AspenCameraHandler: ❌ activeView 是 nil，无法清理")
        }

        AspenCameraViewFactory.viewLock.unlock()

        isPlayerInitialized = false
        result(nil)

        print("AspenCameraHandler: ✅ dispose 完成")
    }

    private func handleSnapshot(result: @escaping FlutterResult) {
        print("AspenCameraHandler: MethodChannel 截图")

        // 通知 PlatformView 执行截图
        // 这里需要通过通知机制传递给当前的 AspenCameraView
        NotificationCenter.default.post(
            name: NSNotification.Name("AspenCameraSnapshotRequest"),
            object: nil,
            userInfo: ["result": result]
        )
    }

    private func handleSwitchQuality(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let quality = args["quality"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGS",
                message: "参数缺失",
                details: nil
            ))
            return
        }

        print("AspenCameraHandler: MethodChannel 切换清晰度: \(quality)")

        // 通知 PlatformView 切换清晰度
        NotificationCenter.default.post(
            name: NSNotification.Name("AspenCameraSwitchQuality"),
            object: nil,
            userInfo: ["quality": quality, "result": result]
        )
    }

    // ═══════════════════════════════════════════════════════════════
    // EventChannel - 推送事件到 Flutter
    // ═══════════════════════════════════════════════════════════════

    /// 发送事件到 Flutter
    ///
    /// - Parameters:
    ///   - type: 事件类型（"p2p" 或 "player"）
    ///   - event: 事件名称（"connected", "video_ready" 等）
    ///   - data: 附加数据（可选）
    func sendEvent(type: String, event: String, data: [String: Any]? = nil) {
        guard let sink = eventSink else {
            print("AspenCameraHandler: EventSink 未初始化，无法发送事件")
            return
        }

        // 必须在主线程推送
        DispatchQueue.main.async {
            var eventData: [String: Any] = [
                "type": type,
                "event": event
            ]
            if let data = data {
                eventData["data"] = data
            }

            print("AspenCameraHandler: EventChannel 推送事件: \(eventData)")
            sink(eventData)
        }
    }

    /// 发送错误到 Flutter
    func sendError(code: String, message: String, details: Any? = nil) {
        guard let sink = eventSink else { return }

        DispatchQueue.main.async {
            sink(FlutterError(code: code, message: message, details: details))
        }
    }

    // ─────────────────────────────────────────────────────────
    // 注销通道
    // ─────────────────────────────────────────────────────────
    func unregister() {
        print("AspenCameraHandler: 注销通信通道")
        methodChannel.setMethodCallHandler(nil)
        eventChannel.setStreamHandler(nil)
    }
}

// ═══════════════════════════════════════════════════════════════
// EventChannel StreamHandler 实现
// ═══════════════════════════════════════════════════════════════
extension AspenCameraHandler: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        print("AspenCameraHandler: Flutter 开始监听事件")
        self.eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        print("AspenCameraHandler: Flutter 取消监听事件")
        self.eventSink = nil
        return nil
    }
}
